src/Domain/Services/CartCheckoutPageFormat.php

```diff
@@ -0,0 +1,365 @@
+		// and it won't match. See block_attributes_regex() above; uses `?:`
+		// to avoid this issue.
+
+		// The target shortcode, with optional whitespace before/after.
+		$shortcode_matcher = '/^' .
+			$optional_whitespace .
+			$shortcode_regex .
+			$optional_whitespace .
+			'$/s';
+
+		// The target shortcode, in a shortcode block, with optional whitespace between elements.
+		$shortcode_block_matcher = '/^' .
+			$optional_whitespace .
+			$shortcode_block_start .
+			$optional_whitespace .
+			$shortcode_regex .
+			$optional_whitespace .
+			$shortcode_block_end .
+			$optional_whitespace .
+			'$/s';
+
+		if ( preg_match( $shortcode_matcher, $page_content ) || preg_match( $shortcode_block_matcher, $page_content ) ) {
+			return 'shortcode';
+		}
+
+		$target_block_start = self::block_start_regex( $block_type );
+		$target_block_end   = self::block_end_regex( $block_type );
+		$wildcard           = '.*';
+
+		// The target block start/end tags, with any content between, and optional whitespace before/after.
+		$block_matcher = '/^' .
+			$optional_whitespace .
+			$target_block_start .
+			$wildcard .
+			$target_block_end .
+			$optional_whitespace .
+			'$/s';
+
+		if ( preg_match( $block_matcher, $page_content ) ) {
+			return 'block';
+		}
+
+		return 'custom';
+	}
+
+	/**
+	 * Get info about what blocks & shortcodes are used on the cart & checkout pages.
+	 *
+	 * Note: this code is adapted from WooCommerce core: https://github.com/woocommerce/woocommerce/blob/60d6510cb049836ae5c92482f0c8853436e42726/includes/class-wc-tracker.php#L755
+	 *
+	 * @return array|boolean Array with page ids and whether cart or checkout
+	 * blocks/shortcodes are present.
+	 */
+	public function get_cart_checkout_status() {
+		$info = [
+			'cart_page_id'                              => 0,
+			'checkout_page_id'                          => 0,
+			'cart_page_format'                          => 'custom',
+			'checkout_page_format'                      => 'custom',
+			// We're also returning these values, similar to the values in
+			// wc-tracker (see doc block above). Before merging this PR,
+			// evaluate if we should share code with core for these values.
+			// This PR uses a more robust & flexible technique (regex + has_shortcode).
+			'cart_page_contains_cart_shortcode'         => false,
+			'checkout_page_contains_checkout_shortcode' => false,
+			'cart_page_contains_cart_block'             => false,
+			'checkout_page_contains_checkout_block'     => false,
+		];
+
+		$cart_page_id     = wc_get_page_id( 'cart' );
+		$checkout_page_id = wc_get_page_id( 'checkout' );
+
+		if ( ! $cart_page_id || ! $checkout_page_id ) {
+			return $info;
+		}
+
+		$info['cart_page_id']     = $cart_page_id;
+		$info['checkout_page_id'] = $checkout_page_id;
+
+		$cart_page     = get_post( $cart_page_id );
+		$checkout_page = get_post( $checkout_page_id );
+
+		if ( ! $cart_page || ! $checkout_page ) {
+			return $info;
```

Comment:
Notice this early return makes it so that if one of the pages is not defined, the format of the other would not be loaded correctly.
![imatge](https://user-images.githubusercontent.com/3616980/96972334-aebcba80-1516-11eb-8b4e-c03b6f4730b8.png)

Would refactoring this to something like the code below work?
```PHP
if ( $cart_page ) {
  // do all Cart page logic here
}
if ( $checkout_page ) {
  // do all Checkout page logic here
}
return $info;
```

Reply:
Yep - good catch - fixed here: 556ec1e6cb07e8640033a36c32976916df415ba5 😄